eth_account/_utils/structured_data/hashing.py

```diff
@@ -0,0 +1,279 @@
+            for i in range(len(array_dimensions)):
+                if len(parsed_type.arrlist[i]) == 0:
+                    # Skip empty or dynamically declared dimensions
+                    continue
+                if array_dimensions[i] != parsed_type.arrlist[i][0]:
+                    # Dimensions should match with declared schema
+                    raise TypeError(
+                        "Array data `{0}` has dimensions `{1}` whereas the "
+                        "schema has dimensions `{2}`".format(
+                            value,
+                            array_dimensions,
+                            tuple(map(lambda x: x[0], parsed_type.arrlist)),
+                        )
+                    )
+
+            array_items = flatten_multidimensional_array(value)
+            array_items_encoding = [
+                encode_data(parsed_type.base, types, array_item)
+                for array_item in array_items
+            ]
+            concatenated_array_encodings = ''.join(array_items_encoding)
+            hashed_value = keccak(concatenated_array_encodings)
+            yield "bytes32", hashed_value
+        else:
+            # First checking to see if type is valid as per abi
+            if not is_valid_abi_type(field["type"]):
+                raise TypeError(
+                    "Received Invalid type `{0}` in the struct `{1}`".format(
+                        field["type"],
+                        primary_type,
+                    )
+                )
+
+            # Next see if the data fits the specified encoding type
+            if is_encodable(field["type"], value):
+                # field["type"] is a valid type and this value corresponds to that type.
+                yield field["type"], value
+            else:
+                raise TypeError(
+                    "Value of `{0}` ({2}) in the struct `{1}` is of the type `{3}`, but expected "
+                    "{4} value".format(
+                        field["name"],
+                        primary_type,
+                        value,
+                        type(value),
+                        field["type"],
+                    )
+                )
+
+
+def encode_data(primaryType, types, data):
+    data_types_and_hashes = _encode_data(primaryType, types, data)
+    data_types, data_hashes = zip(*data_types_and_hashes)
+    return encode_abi(data_types, data_hashes)
+
+
+def hash_struct(structured_json_string_data, is_domain_separator=False):
+    """
+    The structured_json_string_data is expected to have the ``types`` attribute and
+    the ``primaryType``, ``message``, ``domain`` attribute.
+    The ``is_domain_separator`` variable is used to calculate the ``hashStruct`` as
+    part of the ``domainSeparator`` calculation.
+    """
+    structured_data = json.loads(structured_json_string_data)
+    validate_structured_data(structured_data)
+
+    types = structured_data["types"]
+    if is_domain_separator:
```

Comment:
after reading through this one more time I think this function should be decomposed as follows.  My reasoning is that boolean flags like this are an indication that there is some inner functionality that isn't properly separated, thus requiring a feature flag to be able to re-use that functionality.

Here's where I think the lines should be.

1. Outermost API that the user interacts with is responsible for loading the JSON and validating it.  All remaining functions assume inputs are valid.
2. `hash_domain(structured_data)`:
    - simple wrapper for `keccak(encode_data("EIP712Domain", structured_data["types"], structured_data["domain"]))`
3. `hash_message(structured_data)`:
    - simple wrapper for `keccak(encode_data(structured_data["primaryType"], structured_data["types"], structured_data["message"]))`


Probably good to have some high level nice error messages if any of the `"types"` / `"domain"` / `"message"` / `"primaryType"` keys are missing.